### PR TITLE
feat(across): add send from L1 -> Optimism

### PR DESCRIPTION
**Pull request by amateima**

### CODE REVIEW

File: src/components/CoinSelection/CoinSelection.tsx [hunk 1]
```diff
@@ -31,19 +29,32 @@ import {
...
   const [error, setError] = React.useState<Error>();
   const sendState = useAppSelector((state) => state.send);
-
-  const tokenList = TOKENS_LIST[sendState.currentlySelectedFromChain.chainId];
+  const tokenList = useMemo(() => {
+    if (fromChain === ChainId.MAINNET && toChain === ChainId.OPTIMISM) {
+      return TOKENS_LIST[sendState.currentlySelectedFromChain.chainId].slice(1);
+    }
+    return TOKENS_LIST[sendState.currentlySelectedFromChain.chainId];
+  }, [fromChain, toChain, sendState.currentlySelectedFromChain.chainId]);
```
Comment (amateima) [inline src/components/CoinSelection/CoinSelection.tsx:41]: `tokenList` supported by Optimism for deposits is not the same with our static TOKENS_LIST, this is why fetching balances by index doesn't work. I need to keep a map between tokens and the corresponding balance. This fixed my problem, but it has to be stored higher in the state

File: src/components/CoinSelection/CoinSelection.tsx [hunk 2]
```diff
@@ -31,19 +29,32 @@ import {
...
   );
+  const tokenBalanceMap = useMemo(() => {
+    return TOKENS_LIST[fromChain].reduce((acc, val, idx) => {
+      return {
+        ...acc,
+        [val.address]: balances ? balances[idx] : undefined,
+      };
+    }, {} as Record<string, BigNumber | undefined>);
+  }, [balances, fromChain])
+
```
Comment (amateima) [inline src/components/CoinSelection/CoinSelection.tsx:57]: `tokenList` supported by Optimism for deposits is not the same with our static TOKENS_LIST, this is why fetching balances by index doesn't work. I need to keep a map between tokens and the corresponding balance. This fixed my problem, but it has to be stored higher in the state

File: src/state/hooks.ts [hunk 3]
```diff
@@ -336,3 +374,87 @@ export function useBalance(params: {
...
+    }
+  }, [amount, fees, token, isConnected, optimismBridge, signer]);
+
+  const approve = useCallback(() => {
+    if (!signer) return;
+    return optimismBridge.approve(signer, token, MAX_APPROVAL_AMOUNT);
+  }, [optimismBridge, signer, token]);
+
+  const canSend = true;
+  const hasToSwitchChain = false;
```
Comment (daywiss) [inline src/state/hooks.ts:450]: do we need to calculate these?

File: src/state/hooks.ts [hunk 4]
```diff
@@ -336,3 +374,87 @@ export function useBalance(params: {
...
+      owner: account!,
+      spender: bridgeAddress,
+      amount,
+    },
+    { skip: !account || !isConnected || !chainId }
+  );
+  const canApprove = balance.gte(amount) && amount.gte(0);
+  const hasToApprove = allowance?.hasToApprove ?? true;
+  //TODO: Add fees
+  const [fees] = useState({
```
Comment (daywiss) [inline src/state/hooks.ts:410]: need to figure out fees eventually

**Review by daywiss — APPROVED** [review]
looks great as far as i can tell, couple things to figure out here

1. something is going on with commit history
2. need to make a new card to add in estimated gas fees or whatever other fees there are
3. need to disable send when not on chain or amount out of bounds

anyways these can be tackled in future prs, except maybe commit history

File: src/components/AddressSelection/AddressSelection.tsx [hunk 5]
```diff
@@ -121,15 +121,15 @@ const AddressSelection: React.FC = () => {
                 return (
                   <Item
                     className={
-                      t === sendState.currentlySelectedToChain ? "disabled" : ""
+                      t === sendState.currentlySelectedToChain || t.chainId !== ChainId.OPTIMISM ? "disabled" : ""
```
[the remark this answers is not in the document]
Comment (amateima) [reply src/components/AddressSelection/AddressSelection.tsx:124]: Hide just for now Arb and Boba. 

> will this need to change to enable arbitrum, boba, etc?

yes

File: src/state/hooks.ts (hunk 3 above)
Comment (Tulun) [reply src/state/hooks.ts:450]: canSend is apparently quite complicated, so absolutely.

**Comment by amateima** [comment]
> looks great as far as i can tell, couple things to figure out here
> 
> 1. something is going on with commit history
> 2. need to make a new card to add in estimated gas fees or whatever other fees there are
> 3. need to disable send when not on chain or amount out of bounds
> 
> anyways these can be tackled in future prs, except maybe commit history

1. I can rework the commit history, but these commits will be squashed, so I don't think it's a problem.
2. I intended to make a separate PR because it will require changes in multiple places
3. fixed. thx

File: src/state/hooks.ts (hunk 3 above)
Comment (amateima) [reply src/state/hooks.ts]: added 👍

File: src/state/hooks.ts (hunk 4 above)
Comment (amateima) [reply src/state/hooks.ts:411]: I will make a separate PR for this because it will require changes in multiple places

**Review by Tulun — APPROVED** [review]
Think it's ok given you're addressing David's concerns.

**Pull request merged**